Бочаров Александр

Geratoptus opened the pull request.

Geratoptus committed:
ObjectComparison рефактор

Geratoptus committed:
NumberValidatorTests рефактор

Folleach commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -14,16 +15,8 @@ public void CheckCurrentTsar()
-        ClassicAssert.AreEqual(expectedTsar.Parent.Height, actualTsar.Parent.Height);
-        ClassicAssert.AreEqual(expectedTsar.Parent.Parent, actualTsar.Parent.Parent);
+        actualTsar.Should().BeEquivalentTo(expectedTsar, options => 
+            options.Excluding(x => x.Path.EndsWith("Id")));
```
Хорошая идея использовать Fluent Assertions, но что будет, если царю присвоить его родной город? Тип которого будет выглядеть, например так:
```cs
public class Town
{
    public int Id;
    public string Name;
}
```

В таком случае, хочется проверять идентификаторы городов, но сейчас это невозможно

Folleach replied:
Коммит 7dccd562 действительно исправляет ситуацию с городом.  
Можно ли придумать ещё ситуации, в которых тестовый метод поведёт себя неожиданно?

Folleach replied:
Думаю, теперь это точно работает корректно!  
Правда, если я применю рефакторинг к полю `Id`, то мне придётся пойти и исправить все тесты... Потому что в тестах написаны строковые литералы. Может быть есть варианты избежать массового переименования строчек своими руками?

Folleach replied:
resolved

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
 public class NumberValidatorTests
 {
+    [Test, TestCaseSource(nameof(ValidatorTestCases))]
+    public bool IsValid(NumberValidator numberValidator, string number)
```
Нет необходимости помечать метод атрибутом `Test` при наличии `TestCaseSource`  
https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcasesource.html
> The attribute additionally identifies the method as a test method.

Folleach replied:
resolved

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
+    }
+
     [Test]
     public void Test()
```
Из названия метода не понятно, что проверяет этот тест  
Предлагаю для именования использовать AAA паттерн

Folleach replied:
`TestName` свойство перекрывает название метода
![image](https://github.com/user-attachments/assets/48e21ffd-d676-46d0-97e1-88cf90f2b16b)

Но в целом хорошо, пусть будет так
resolved

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
+        {
+            yield return new TestCaseData(new NumberValidator(17, 2, true), "0.0")
+                .SetName("precision=17, scale=2, onlyPositive=true, number=0.0")
+                .Returns(true);
```
Использовать генератор - хорошая идея для передачи сложных параметров в тестовый метод или для создания перестановок:
- `(false, 1, 2, 3)`
- `(true, 1, 2, 3)`

Но в текущем случае, можно пойти более простым путём. Обрати внимание на эту страницу документации NUnit: https://docs.nunit.org/articles/nunit/writing-tests/attributes/testcase.html

Какие видишь проблемы в текущем использовании `TestCaseDate`?

Folleach replied:
resolved

Folleach replied:
Проблема с `TestCaseDate` перенеслась на `TestCase` :)

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
+            
+            yield return new TestCaseData(new NumberValidator(3, 2, true), "+1.23")
+                .SetName("precision=3, scale=2, onlyPositive=true, number=+1.23")
+                .Returns(false);
```
Чем этот тест отличается от теста выше?
```cs
yield return new TestCaseData(new NumberValidator(3, 2, true), "+0.00")
    .SetName("precision=3, scale=2, onlyPositive=true, number=+0.00")
    .Returns(false);
```

Стоит избегать проверки одинаковой логики, так как это лишь добавляет объём к тестам, не проверяя дополнительных кравевых случаев.  
Уберёшь повторения тестов?

Если хочется тестировать, как поведёт себя программа/метод при передачи разнообразных чисел, можно посмотреть в сторону fuzzing тестов  
https://en.wikipedia.org/wiki/Fuzzing  
Но в данном случае, можно обойтись без них

Folleach replied:
Наверное не стоило так много варианций с `,.` делать, но в целом меня устраивает  

resolved

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
+        var rightCreation = () => new NumberValidator(1, 0, true);
+        
+        wrongCreation.Should().Throw<ArgumentException>();
+        rightCreation.Should().NotThrow<ArgumentException>();
```
Падение теста `wrongCreation` останавливает тест `rightCreation`

Folleach replied:
resolved

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
-        ClassicAssert.IsFalse(new NumberValidator(3, 2, true).IsValidNumber("-1.23"));
-        ClassicAssert.IsFalse(new NumberValidator(3, 2, true).IsValidNumber("a.sd"));
+        var wrongCreation = () => new NumberValidator(-1, 2, true);
+        var rightCreation = () => new NumberValidator(1, 0, true);
```
Кажется, можно найти ещё краевых случаев в конструкторе `NumberValidation`

Folleach replied:
resolved

Folleach commented on Testing/Basic/Homework/1. ObjectComparison/ObjectComparison.cs:
```diff
@@ -35,6 +28,19 @@ public void CheckCurrentTsar_WithCustomEquality()
             new Person("Vasili III of Russia", 28, 170, 60, null));
 
         // Какие недостатки у такого подхода? 
+        /*
+         * 1. Метод AreEqual проверяет только определенные свойства класса, поэтому если
+         * в будущем придется добавить новые или удалить текущие, то их также придется
+         * добавить/удалить в этом методе.
+         * 2. Написание собственного метода для проверки равенства теоретически усложнит
+         * понимание теста, так как нужно разобраться в работе написанного метода. При
+         * этом можно было спокойно воспользоваться стандартными методами.
+         * 3. Так как метод возвращает bool, то и сравнивать результат этого метода
+         * приходится с True/False, поэтому при провале теста мы не получаем
+         * конкретной информации в чем была проблема. В отличие от первого теста с
+         * применением AreEqual для каждого поля в отдельности. (если используем
+         * FluentAssertions, то вообще сказка)
+         */
```
![image](https://github.com/user-attachments/assets/d12a276f-886b-44f5-8b5b-9d902baf4205)

Folleach commented on Testing/Basic/Homework/2. NumberValidator/NumberValidatorTests.cs:
```diff
@@ -7,25 +8,85 @@ namespace HomeExercise.Tasks.NumberValidator;
+        return numberValidator.IsValidNumber(number);
+    }
+    
+    public static IEnumerable ValidatorTestCases
```
Я нашёл ещё 3 случая, которые написаны в `NumberValidator`, но не обработаны в тесте.

<details>

<summary><b>Подсказка</b></summary>

Регулярные выражения можно представить в виде конечного автомата, это может облегчить задачу.

```mermaid
graph LR
  START
  END
  MINUS[​-]
  PLUS[​+]
  DIGIT1[digit]
  DOT[.]
  COMMA[,]
  DIGIT2[digit]
  
  START --> MINUS
  START --> PLUS
  START --> DIGIT1
%%  DIGIT1 --> DIGIT1
  MINUS --> DIGIT1
  PLUS --> DIGIT1
  DIGIT1 --> DOT
  DIGIT1 --> COMMA
  DOT --> DIGIT2
  COMMA --> DIGIT2
%%  DIGIT2 --> DIGIT2
  DIGIT1 --> END
  DIGIT2 --> END
```

</details>

Folleach replied:
Прикольный случай с `1.1.1`, даже не подумал о нём )

А вот я могу сейчас сделать патч в метод `IsValidNumber`, твои тесты все пройдут, но логика из комментария будет нарушена.

Folleach replied:
resolved

Folleach commented:
### Просто рекомендация к PR
**Это не нужно править**

Сейчас у тебя Pull Request создан из ветки `master` твоего репозитория, в целом допустимо, но тем самым могут возникнуть проблемы при сихронизации ветки `master`, когда будешь делать pull из основного репозитория.  
Если в основном репозитории появились новые коммиты, то в твоей ветке `master` будет создан merge commit, который появится в этом PR. Таким образом, пока активен PR, не будет возможности что-то сделать с `master` веткой форка.  

Из этих соображений, лучше создавать в форке отдельную ветку для Pull Request. Это позволит легко вносить изменения в PR и не получать нежелательных конфликтов. Будут только желательные конфликты, хех :)

Geratoptus committed:
ObjectComparison теперь при сравнении исключаются только Id людей.

Geratoptus committed:
NumberValidatorTests рефактор тестов конструктора NumberValidator

Geratoptus committed:
NumberValidatorTests рефактор теста для метода IsValidNumber:
- Параметризация через TestCase
- Убраны повторяющиеся тесты
- Добавлены недостающие тесты

Geratoptus committed:
NumberValidatorTests тест для метода IsValidNumber - убрал TestName'ы

Geratoptus committed:
ObjectComparison теперь при сравнении исключает только поле .Id, а не все, что оканчивается на Id

Geratoptus committed:
NumberValidatorTests добавка краевых случаев

Geratoptus committed:
ObjectComparison теперь имя поля, соответствующего Id автоматически изменяется при рефакторинге
